parsec/backend/postgresql/handler.py

```diff
@@ -1,52 +1,121 @@
 
 
-async def init_db(url: str) -> None:
+async def migrate_db(url: str, migration_files: list, dry_run: bool) -> None:
     """
-    Returns: if the database was already initialized
+    Returns: A result dict
     Raises:
         triopg.exceptions.PostgresError
     """
     async with triopg.connect(url) as conn:
-        return await _init_db(conn)
-
+        return await _migrate_db(conn, migration_files, dry_run)
+
+
+async def _migrate_db(conn, migration_files, dry_run):
+
+    errors = []
+    allready_applied = []
+    new_apply = []
+    to_apply = []
+    rows = []
+
+    idx_limit, error = await _idx_limit(conn, migration_files)
+    if not error:
+        for idx, migration in enumerate(migration_files, 1):
+            if idx <= idx_limit:
+                allready_applied.append(migration)
+            else:
+                if not dry_run:
+                    error = await _apply_migration(conn, migration)
+                    if error:
+                        errors.append(error)
+                        break
+                    else:
+                        new_apply.append(migration)
+                        rows.append((migration, datetime.datetime.utcnow()))
+                else:
+                    to_apply.append(migration)
+
+        if rows:
+            statement = "INSERT INTO migration (name, applied) VALUES ($1, $2)"
+            await conn.executemany(statement, rows)
+
+    return collections.OrderedDict(
+        (
+            ("already_applied", allready_applied),
+            ("new_apply", new_apply),
+            ("to_apply", to_apply),
+            ("errors", errors),
+        )
+    )
 
-async def _init_db(conn):
-    if await _is_db_initialized(conn):
-        return True
 
+async def _apply_migration(conn, migration_file):
     async with conn.transaction():
-        query = read_text(__package__, f"init_tables.sql")
-        await conn.execute(query)
-    return False
-
-
-async def _is_db_initialized(conn):
-    # TODO: not a really elegant way to determine if the database
-    # is already initialized...
-    root_key = await conn.fetchrow(
-        """
-        SELECT true FROM pg_catalog.pg_tables WHERE tablename = 'user_';
-        """
-    )
-    return bool(root_key)
+        query = read_text(migrations, f"{migration_file}")
+        error = None
+        if not query:
+            error = (migration_file, "Empty migration file")
+        else:
+            try:
+                await conn.execute(query)
+            except PostgresError as e:
+                error = (migration_file, e.message)
+    return error
+
+
+async def _last_migration_row(conn):
+    query = """
+        SELECT name FROM migration ORDER BY applied desc LIMIT 1
+    """
+    return await conn.fetchval(query)
+
+
+async def _idx_limit(conn, migration_files):
+    idx_limit = None
+    error = None
+    try:
+        last_migration_row = await _last_migration_row(conn)
+    except UndefinedTableError:
```

Comment:
We should add here a comment about the fact this table was added by migration 2

Reply:
Done